File: src/services/recording.service.ts
import { invoke } from "@tauri-apps/api/core"

export type RecordingStatus = "idle" | "recording" | "paused"

const isTauri = (): boolean => {
  return (
    typeof window !== "undefined" &&
    ("__TAURI__" in window || "__TAURI_INTERNALS__" in window)
  )
}

export interface RecordingMetadata {
  output_path: string
  sample_rate: number
  channels: number
  started_at: string
  finished_at: string | null
}

export interface RecordingInfo {
  status: RecordingStatus
  metadata: RecordingMetadata | null
}

export const recordingService = {
  async start(outputPath: string): Promise<RecordingMetadata> {
    if (!isTauri()) {
      throw new Error("Recording is only available in the Tauri app")
    }
    return invoke<RecordingMetadata>("start_recording", {
      outputPath,
    })
  },

  async stop(): Promise<RecordingMetadata> {
    if (!isTauri()) {
      throw new Error("Recording is only available in the Tauri app")
    }
    return invoke<RecordingMetadata>("stop_recording")
  },

  async pause(): Promise<string> {
    if (!isTauri()) {
      throw new Error("Recording is only available in the Tauri app")
    }
    return invoke<string>("pause_recording")
  },

  async resume(): Promise<string> {
    if (!isTauri()) {
      throw new Error("Recording is only available in the Tauri app")
    }
    return invoke<string>("resume_recording")
  },

  async getStatus(): Promise<RecordingInfo> {
    if (!isTauri()) {
      return { status: "idle", metadata: null }
    }
    return invoke<RecordingInfo>("get_recording_status")
  },
}

File: src-tauri/src/domain/audio/services/recording_service.rs
use crate::domain::audio::entities::recording::{RecordingConfig, RecordingInfo, RecordingMetadata};
use crate::domain::shared::DomainResult;

pub trait RecordingService: Send + Sync {
    fn start_recording(&mut self, config: RecordingConfig) -> DomainResult<RecordingMetadata>;
    fn stop_recording(&mut self) -> DomainResult<RecordingMetadata>;
    fn pause_recording(&mut self) -> DomainResult<()>;
    fn resume_recording(&mut self) -> DomainResult<()>;
    fn get_recording_info(&self) -> RecordingInfo;
}
